### PR TITLE
Add missing tests to test_dali_variable_batch_size.py

### PR DESCRIPTION
**Pull request by JanuszL**
- add tests for missing operators in test_dali_variable_batch_size.py,
  transforms, math, segmentation to name a few

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds missing tests to test_dali_variable_batch_size.py

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     add tests for missing operators in test_dali_variable_batch_size.py,  transforms, math, segmentation to name a few
 - Affected modules and functionalities:
     test_dali_variable_batch_size.py,
 - Key points relevant for the review:
     test code
 - Validation and testing:
     CI
 - Documentation (including examples):
     NA

**JIRA TASK**: *[Use DALI-2102]*

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Review by klecki — COMMENTED** [review]
Posting comments, as there is one about the OF-related stuff.

File: dali/test/python/test_dali_variable_batch_size.py [hunk 1]
```diff
@@ -372,18 +445,286 @@ def pipe_no_input(max_batch_size, input_data, device):
...
+    ]
+    run_pipeline(get_data, pipeline_fn=pipe, devices=['cpu'])
+
+def test_math_ops():
+    def pipe(max_batch_size, input_data, device):
+        pipe = Pipeline(batch_size=max_batch_size, num_threads=4, device_id=0)
+        with pipe:
+            # just to drive the variable batch size.
+            data = fn.external_source(source=input_data, cycle=False, device=device)
+            processed = [data * 2,
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py]: Done (I think)

File: dali/test/python/test_dali_variable_batch_size.py [hunk 2]
```diff
@@ -372,18 +445,286 @@ def pipe_no_input(max_batch_size, input_data, device):
...
+                 dmath.pow(data, 2),
+                 dmath.fpow(data, 1.5),
+                 dmath.min(data, 2),
+                 dmath.max(data, 50),
+                 dmath.clamp(data, 10, 50),
+                 dmath.sqrt(data),
+                 dmath.rsqrt(data),
+                 dmath.cbrt(data),
+                 dmath.exp(data),
+                 dmath.exp(data),
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py]: Fixed

File: dali/test/python/test_dali_variable_batch_size.py [hunk 3]
```diff
@@ -60,6 +66,26 @@
...
+    try:
+        import pynvml
+        pynvml.nvmlInit()
+        handle = pynvml.nvmlDeviceGetHandleByIndex(device_id)
+        compute_cap = pynvml.nvmlDeviceGetCudaComputeCapability(handle)
+        compute_cap = compute_cap[0] + compute_cap[1] / 10.
+    except ModuleNotFoundError:
+        pass
+
+    is_gds_supported_var = compute_cap >= 7.5
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py]: Fixed

File: dali/test/python/test_dali_variable_batch_size.py [hunk 4]
```diff
@@ -372,18 +445,294 @@ def pipe_no_input(max_batch_size, input_data, device):
...
+            out = fn.squeeze(out, axis_names="Z")
+        pipe.set_outputs(out)
+        return pipe
+
+    check_pipeline(generate_data(31, 13, image_like_shape_generator, lo=0, hi=255, dtype=np.uint8),
+                   pipeline_fn=pipe, input_layout="HWC")
+
+
+def test_box_encoder_op():
+    def coco_anchors():
```
Comment (klecki) [inline dali/test/python/test_dali_variable_batch_size.py:620]: How about importing it from the coco test rather than having 3 copies of the same piece of code?

File: dali/test/python/test_dali_variable_batch_size.py [hunk 5]
```diff
@@ -372,18 +445,294 @@ def pipe_no_input(max_batch_size, input_data, device):
...
+            boxes, lables = fn.external_source(device=device, source=input_data, num_outputs=2)
+            processed, _ = fn.box_encoder(boxes, lables, anchors=coco_anchors())
+        pipe.set_outputs(processed)
+        return pipe
+
+    def get_data(batch_size):
+        obj_num = random.randint(1, 20)
+        test_box_shape = [obj_num, 4]
+        test_lables_shape = [obj_num, 1]
+        bboxes = [(np.random.randint(0, 255, size=test_box_shape, dtype=np.uint8) / 255).astype(dtype=np.float32) for _ in range(batch_size)]
```
Comment (klecki) [inline dali/test/python/test_dali_variable_batch_size.py:681]: Nitpick:
```suggestion
        bboxes = [np.random.random(size=test_box_shape, dtype=np.uint8) .astype(dtype=np.float32) for _ in range(batch_size)]
```
Just floats in [0, 1]?

And a question: does it work ok with invalid boxes?

File: dali/test/python/test_dali_variable_batch_size.py [hunk 6]
```diff
@@ -697,6 +1027,43 @@ def pipe(max_batch_size, input_data, device, input_layout):
     check_pipeline(generate_data(31, 13, (5, 160, 80, 3), lo=0, hi=255, dtype=np.uint8),
                    pipeline_fn=pipe, input_layout="FHWC")
 
+def test_segmentation_select_masks():
+    def get_data_source(*args, **kwargs):
+        return lambda: make_batch_select_masks(*args, **kwargs)
```
Comment (klecki) [inline dali/test/python/test_dali_variable_batch_size.py:1032]: Why the lambda, can't you call it directly?

**Review by klecki — COMMENTED** [review]
Mostly small nitpicks, I assume most is based on the exisiting tests and looks ok.

File: dali/test/python/test_dali_variable_batch_size.py [hunk 7]
```diff
@@ -372,18 +445,294 @@ def pipe_no_input(max_batch_size, input_data, device):
...
+        out = [dl_tensor_operation(t) for t in tensors]
+        return out
+
+    def pipe(max_batch_size, input_data, device, input_layout=None):
+        pipe = Pipeline(batch_size=max_batch_size, num_threads=4, device_id=0,
+                        exec_async=False, exec_pipelined=False)
+        with pipe:
+            input = fn.external_source(source=input_data, cycle=False, device=device,
+                                       layout=input_layout)
+            output = fn.dl_tensor_python_function(input, function=batch_dl_tensor_operation, batch_processing=True)
```
Comment (klecki) [inline dali/test/python/test_dali_variable_batch_size.py:506]: Would it make sense to test non-batch variant?

File: dali/test/python/test_dali_variable_batch_size.py (hunk 6 above)
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py:1032]: Done

File: dali/test/python/test_dali_variable_batch_size.py (hunk 5 above)
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py:681]: Done
No idea, but I don't think this is the place to test this aspect of bbox encoder.

File: dali/test/python/test_dali_variable_batch_size.py (hunk 4 above)
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py:620]: Done

File: dali/test/python/test_dali_variable_batch_size.py [hunk 8]
```diff
@@ -139,25 +184,36 @@ def check_pipeline(input_epoch, pipeline_fn, *, devices: list = ['cpu', 'gpu'],
...
+                        and each item is a single batch of
     :param pipeline_fn: Function, that returns created (but not built) pipeline.
                         Its signature should be (at least):
                         pipeline_fn(max_batch_size, input_data, device, ...)
     :param devices: Devices to run the check on
     :param eps: Epsilon for mean error
     :param pipeline_fn_args: Additional args to pipeline_fn
     """
+    def get_batch_size(batch):
+        if isinstance(batch, Iterable) and not types._is_numpy_array(batch):
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py:199]: Done

File: dali/test/python/test_dali_variable_batch_size.py [hunk 9]
```diff
@@ -139,25 +184,36 @@ def check_pipeline(input_epoch, pipeline_fn, *, devices: list = ['cpu', 'gpu'],
 
     This function conducts qualitative verification. It compares the result of
     running multiple iterations of the same pipeline (with possible varying batch sizes,
-    accoring to `input_epoch`) with results of the ad-hoc created pipelines per iteration
+    according to `input_epoch`) with results of the ad-hoc created pipelines per iteration
 
-    :param input_epoch: List of numpy arrays, where every item is a single batch
+    :param input_epoch: List where every item is numpy arrays or list of numpy arrays,
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py:189]: Done

File: dali/test/python/test_dali_variable_batch_size.py [hunk 10]
```diff
@@ -123,9 +158,19 @@ def run_pipeline(input_epoch, pipeline_fn, *, devices: list = ['cpu', 'gpu'], **
     :param devices: Devices to run the check on
     :param pipeline_fn_args: Additional args to pipeline_fn
     """
+    def get_batch_size(batch):
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py:161]: Done

File: dali/test/python/test_dali_variable_batch_size.py (hunk 7 above)
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py]: Done

File: dali/test/python/test_dali_variable_batch_size.py [hunk 11]
```diff
@@ -60,6 +66,26 @@
 """
 
 
+is_of_supported_var = None
+def is_of_supported(device_id=0):
+    global is_of_supported_var
+    if is_of_supported_var is not None:
```
Comment (szalpal) [inline dali/test/python/test_dali_variable_batch_size.py:72]: This way, we would only occasionally test the `fn.optical_flow`, is that correct? I.e. only if we get lucky and get the machine with OF supported.

I think resolving this is rather out of scope of this PR though

File: dali/test/python/test_dali_variable_batch_size.py [hunk 12]
```diff
@@ -139,25 +185,36 @@ def check_pipeline(input_epoch, pipeline_fn, *, devices: list = ['cpu', 'gpu'],
...
     :param pipeline_fn: Function, that returns created (but not built) pipeline.
                         Its signature should be (at least):
                         pipeline_fn(max_batch_size, input_data, device, ...)
     :param devices: Devices to run the check on
     :param eps: Epsilon for mean error
     :param pipeline_fn_args: Additional args to pipeline_fn
     """
+    def get_batch_size(batch):
+        if isinstance(batch, tuple):
+            return get_batch_size(batch[0])
```
Comment (szalpal) [inline dali/test/python/test_dali_variable_batch_size.py:202]: Is this function the same in `run_pipeline` and `check_pipeline`? If so, maybe extracting to `_get_batch_size`?

File: dali/test/python/test_dali_variable_batch_size.py [hunk 13]
```diff
@@ -98,13 +124,22 @@ def generate_data(max_batch_size, n_iter, sample_shape, lo=0., hi=1., dtype=np.f
...
+            output = input
+        else:
+            if needs_input:
+                output = operator_fn(input, **opfn_args)
+            else:
+                output = operator_fn(**opfn_args)
+        if needs_input:
+            pipe.set_outputs(output)
+        else:
+            pipe.set_outputs(output, input)
```
Comment (szalpal) [inline dali/test/python/test_dali_variable_batch_size.py:142]: This is a hack, so that the `external_source` won't be pruned, correct? If so, maybe a comment here would be good?

File: dali/test/python/test_dali_variable_batch_size.py [hunk 14]
```diff
@@ -263,6 +329,15 @@ def test_ops_image_default_args():
     (fn.rotate, {'angle': 25}),
     (fn.transpose, {'perm': [2, 0, 1]}),
     (fn.warp_affine, {'matrix': (.1, .9, 10, .8, -.2, -20)}),
+    (fn.expand_dims, {'axes': 1, 'new_axis_names': "Z"}),
+    (fn.grid_mask, {'tile': 51, 'ratio': 0.38158387, 'angle': 2.6810782}),
+    (numba_function, {'run_fn': numba_set_all_values_to_255_batch, 'out_types': [types.UINT8],
```
Comment (szalpal) [inline dali/test/python/test_dali_variable_batch_size.py:334]: Apologises for nitpicking, but maybe keeping the alphabetic order of `ops_image_custom_args` would look nice :)

File: dali/test/python/test_dali_variable_batch_size.py [hunk 15]
```diff
@@ -289,6 +364,8 @@ def test_float_array_ops():
     (fn.noise.gaussian, {}),
     (fn.noise.shot, {}),
     (fn.noise.salt_and_pepper, {'devices': ['cpu']}),  # TODO(janton): Enalble GPU once salt_and_pepper supports it.
+    (fn.segmentation.random_mask_pixel, {'devices': ['cpu']}),
+    (fn.roi_random_crop, {'crop_shape': [10, 15, 3], 'roi_start': [25, 20, 0], 'roi_shape': [40, 30, 3], 'devices': ['cpu']})
```
Comment (szalpal) [inline dali/test/python/test_dali_variable_batch_size.py:368]: Same nitpick as above. Feel free to ignore it if you think otherwise

File: dali/test/python/test_dali_variable_batch_size.py [hunk 16]
```diff
@@ -697,6 +983,43 @@ def pipe(max_batch_size, input_data, device, input_layout):
...
+        pipe = Pipeline(batch_size=max_batch_size, num_threads=4, device_id=0)
+        with pipe:
+            data = fn.external_source(device=device, source=input_data, cycle=False, layout=input_layout)
+            processed = fn.optical_flow(data, device=device, output_format=4)
+        pipe.set_outputs(processed)
+        return pipe
+
+    check_pipeline(
+        generate_data(31, 13, custom_shape_generator(3, 7, 100, 100, 200, 200, 3, 3), lo=0, hi=255,
+                      dtype=np.uint8),
```
Comment (szalpal) [inline dali/test/python/test_dali_variable_batch_size.py:1019]: I'm wondering, if testing the optical flow with random data is proper. On one hand, we don't really care about correctness, since we just compare pipelines. On the other hand, if the data is random, the OF operator might just output big tensor of zeros... And this might not be the good result to compare between two pipelines.

Just a food for thought

**Review by szalpal — APPROVED** [review]

File: dali/test/python/test_dali_variable_batch_size.py (hunk 14 above)
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py]: Done

File: dali/test/python/test_dali_variable_batch_size.py (hunk 12 above)
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py]: Done

File: dali/test/python/test_dali_variable_batch_size.py (hunk 15 above)
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py]: Done

File: dali/test/python/test_dali_variable_batch_size.py (hunk 16 above)
Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py:1019]: Copied data loading from the optical flow test.

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by klecki — APPROVED** [review]
